From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Oreon <oreon@local>
Date: Mon, 20 Apr 2026 18:31:30 -0700
Subject: [PATCH] Oreon: pidl Python ignore/skip, relax NDR client errors, split source3 IDL python

Fix pidl FTBFS on newer Perl (smbXsrv [ignore], winbind client stubs).
Run open_files.idl without --python so pidl does not emit Python glue for
struct timeval and struct timespec (Python.pm cannot map those types).
Keep smbXsrv.idl and the remaining source3 IDL lists on --python for
py_smbXsrv.c and friends. Keep irpc.idl --python in source4 as before.

---
 pidl/lib/Parse/Pidl/Samba4/NDR/Client.pm | 2 +-
 pidl/lib/Parse/Pidl/Samba4/Python.pm     | 6 ++++++
 source3/librpc/idl/wscript_build         | 10 ++++++++++
 3 files changed, 18 insertions(+), 3 deletions(-)

diff --git a/pidl/lib/Parse/Pidl/Samba4/NDR/Client.pm b/pidl/lib/Parse/Pidl/Samba4/NDR/Client.pm
index 84e2beb..4418137 100644
--- a/pidl/lib/Parse/Pidl/Samba4/NDR/Client.pm
+++ b/pidl/lib/Parse/Pidl/Samba4/NDR/Client.pm
@@ -793,7 +793,7 @@ sub ParseFunction($$$)
 		$self->pidl_both(" */");
 		$self->pidl_both("");
 
-		error($e->{ORIGINAL}, "$fn->{NAME}: [out] argument '$e->{NAME}' $reason, skip client functions");
+		warning($e->{ORIGINAL}, "$fn->{NAME}: [out] argument '$e->{NAME}' $reason, skip client functions");
 		return;
 	}
 
diff --git a/pidl/lib/Parse/Pidl/Samba4/Python.pm b/pidl/lib/Parse/Pidl/Samba4/Python.pm
index b342704..3b3898e 100644
--- a/pidl/lib/Parse/Pidl/Samba4/Python.pm
+++ b/pidl/lib/Parse/Pidl/Samba4/Python.pm
@@ -165,6 +165,10 @@ sub FromPythonToUnionFunction($$$$$)
 sub PythonElementGetSet($$$$$$) {
 	my ($self, $name, $cname, $ename, $e, $env) = @_;
 
+	if (has_property($e, "ignore") or has_property($e, "skip")) {
+		return;
+	}
+
 	my $varname = "object->$ename";
 	$self->pidl("static PyObject *py_$name\_get_$e->{NAME}(PyObject *obj, void *closure)");
 	$self->pidl("{");
@@ -215,6 +219,7 @@ sub PythonStruct($$$$$$)
 
 	if ($#{$d->{ELEMENTS}} > -1) {
 		foreach my $e (@{$d->{ELEMENTS}}) {
+			next if (has_property($e, "ignore") or has_property($e, "skip"));
 			$self->PythonElementGetSet($name, $cname, $e->{NAME}, $e, $env);
 		}
 
@@ -222,6 +227,7 @@ sub PythonStruct($$$$$$)
 		$self->pidl("static PyGetSetDef ".$getsetters."[] = {");
 		$self->indent;
 		foreach my $e (@{$d->{ELEMENTS}}) {
+			next if (has_property($e, "ignore") or has_property($e, "skip"));
 			my $etype = "";
 			if (ref($e->{TYPE}) eq "HASH") {
 				$etype = $e->{TYPE}->{NAME};
diff --git a/source3/librpc/idl/wscript_build b/source3/librpc/idl/wscript_build
index 3602ba9..d0e5e96 100644
--- a/source3/librpc/idl/wscript_build
+++ b/source3/librpc/idl/wscript_build
@@ -6,8 +6,18 @@
 
 bld.SAMBA_PIDL_LIST('PIDL',
                     '''open_files.idl
-                       perfcount.idl secrets.idl
-                       smbXsrv.idl
+                    ''',
+                    options='--includedir=%s --header --ndr-parser --client' % topinclude,
+                    output_dir='../gen_ndr')
+
+bld.SAMBA_PIDL_LIST('PIDL',
+                    '''smbXsrv.idl
+                    ''',
+                    options='--includedir=%s --header --ndr-parser --client --python' % topinclude,
+                    output_dir='../gen_ndr')
+
+bld.SAMBA_PIDL_LIST('PIDL',
+                    '''perfcount.idl secrets.idl
                        leases_db.idl
                        rpcd_witness.idl
                     ''',
                     options='--includedir=%s --header --ndr-parser --client --python' % topinclude,
                     output_dir='../gen_ndr')
